Map range operations for NpgsqlRange<T>

austindrenski opened the pull request:
# Map range operations for `NpgsqlRange<T>`
Work in progress. See #311 for more information.

## Implemented:
Adds extension methods with PostgreSQL translation:
```c# 
bool Contains<T>(NpgsqlRange<T> range, T value) where T : IComparable<T> {...}
// WHERE range @> value

bool Contains<T>(NpgsqlRange<T> range, NpgsqlRange<T> value) where T : IComparable<T> {...}
// WHERE range @> value

bool ContainedBy<T>(NpgsqlRange<T> value, NpgsqlRange<T> range) where T : IComparable<T> {...}
// WHERE value <@ range

bool Overlaps<T>(NpgsqlRange<T> a, NpgsqlRange<T> b) where T : IComparable<T> {...}
// WHERE a && b

bool IsStrictlyLeftOf<T>(NpgsqlRange<T> a, NpgsqlRange<T> b) where T : IComparable<T> {...}
// WHERE a << b

bool IsStrictlyRightOf<T>(NpgsqlRange<T> a, NpgsqlRange<T> b) where T : IComparable<T> {...}
// WHERE a >> b

bool DoesNotExtendRightOf<T>(NpgsqlRange<T> a, NpgsqlRange<T> b) where T : IComparable<T> {...}
// WHERE a &< b

bool DoesNotExtendLeftOf<T>(NpgsqlRange<T> a, NpgsqlRange<T> b) where T : IComparable<T> {...}
// WHERE a &> b

bool IsAdjacentTo<T>(NpgsqlRange<T> a, NpgsqlRange<T> b) where T : IComparable<T> {...}
// WHERE a -|- b

NpgsqlRange<T> Union<T>(NpgsqlRange<T> a, NpgsqlRange<T> b) where T : IComparable<T> {...}
// WHERE a + b

NpgsqlRange<T> Intersect<T>(NpgsqlRange<T> a, NpgsqlRange<T> b) where T : IComparable<T> {...}
// WHERE a * b

NpgsqlRange<T> Except<T>(NpgsqlRange<T> a, NpgsqlRange<T> b) where T : IComparable<T> {...}
// WHERE a - b
```
## Planned:
- Implement in-memory functionality for range operators:
  - https://www.postgresql.org/docs/current/static/functions-range.html
- Add unit test coverage
- Look at whether properties on `NpgsqlRange<T>` translate into their Postgres functions.
  - e.g.  `range.LowerBound` -> `lower(range)`
## Issues:
 - ~~`range @> range` generates SQL like `range @> range = TRUE`~~
   - Known: aspnet/EntityFrameworkCore#9143
 - `NpgsqlRange<T>` should implement `IComparable<T>`
   - This is _opt-in_ on the part of Npgsql. If/when implemented, the translation support exists.
 - The equality methods on `NpgsqlRange<T>` are returning unexpected results.

austindrenski commented on src/EFCore.PG/EFCore.PG.csproj:
```diff
@@ -1,5 +1,4 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
-
   <PropertyGroup>
     <VersionPrefix>2.1.0-preview2</VersionPrefix>
```
IDE settings gone awry...fixed

austindrenski commented on src/EFCore.PG/Query/ExpressionTranslators/Internal/NpgsqlRangeOperatorTranslator.cs:
```diff
@@ -0,0 +1,309 @@
+        /// <summary>
+        /// Caches runtime method information for <see cref="NpgsqlRangeExtensions.Equal{T}(NpgsqlRange{T}, NpgsqlRange{T})"/>.
+        /// </summary>
+        private static readonly MethodInfo Equal;
```
Removed default modifiers

austindrenski commented on test/EFCore.PG.FunctionalTests/EFCore.PG.FunctionalTests.csproj:
```diff
@@ -18,17 +16,14 @@
   </ItemGroup>
-
-</Project>
+</Project>
```
IDE settings gone awry...fixed

austindrenski commented on src/EFCore.PG/NpgsqlRangeExtensions.cs:
```diff
@@ -0,0 +1,488 @@
+        [Pure]
+        public static bool Equal<T>(this NpgsqlRange<T> a, NpgsqlRange<T> b) where T : IComparable<T>
+        {
+            throw new NotImplementedException();
```
Single-line methods converted

austindrenski commented on src/EFCore.PG/NpgsqlRangeExtensions.cs:
```diff
@@ -0,0 +1,488 @@
+        /// <value>true</value> if the ranges are not equal; otherwise, <value>false</value>.
+        /// </returns>
+        [Pure]
+        public static bool NotEqual<T>(this NpgsqlRange<T> a, NpgsqlRange<T> b) where T : IComparable<T>
```
Single-line methods converted

austindrenski commented on src/EFCore.PG/NpgsqlRangeExtensions.cs:
```diff
@@ -0,0 +1,488 @@
+        /// <value>true</value> if the first range is less than the second; otherwise, <value>false</value>.
+        /// </returns>
+        [Pure]
+        public static bool LessThan<T>(this NpgsqlRange<T> a, NpgsqlRange<T> b) where T : IComparable<T>
```
I've since removed all of the IComparable and IEquatable extensions. It now intercepts the standard ==/Equals calls

austindrenski commented on src/EFCore.PG/NpgsqlRangeExtensions.cs:
```diff
@@ -0,0 +1,488 @@
+        public static bool Contains<T>(this NpgsqlRange<T> range, T value) where T : IComparable<T>
+        {
+            if (range.IsEmpty)
+            {
```
I've removed most of them. Happy to remove any others I missed. (I tend not to notice them, so there could be more...)

austindrenski commented on src/EFCore.PG/NpgsqlRangeExtensions.cs:
```diff
@@ -0,0 +1,488 @@
+            }
+
+            Comparer<T> comparer = Comparer<T>.Default;
+            int compareLower = comparer.Compare(value, range.LowerBound);
```
I'll give these implementations a more thorough look, pretty sure they fail a few edge cases (in addition to outright bugs)

austindrenski commented on src/EFCore.PG/NpgsqlRangeExtensions.cs:
```diff
@@ -0,0 +1,488 @@
+            int compareUpper = comparer.Compare(value, range.UpperBound);
+
+            bool testLower = compareLower > 0 || compareLower == 0 && range.LowerBoundIsInclusive;
+            bool testUpper = compareUpper > 0 || compareUpper == 0 && range.UpperBoundIsInclusive;
```
I'll give these implementations a more thorough look, pretty sure they fail a few edge cases (in addition to outright bugs)

austindrenski commented on src/EFCore.PG/NpgsqlRangeExtensions.cs:
```diff
@@ -0,0 +1,488 @@
+            }
+
+            Comparer<T> comparer = Comparer<T>.Default;
+            int compareLower = comparer.Compare(value.LowerBound, range.LowerBound);
```
Same as above -- will review

austindrenski commented on src/EFCore.PG/NpgsqlRangeExtensions.cs:
```diff
@@ -0,0 +1,488 @@
+        /// <value>true</value> if the range contains the specified value; otherwise, <value>false</value>.
+        /// </returns>
+        [Pure]
+        public static bool ContainedBy<T>(this T value, NpgsqlRange<T> range) where T : IComparable<T>
```
Great point. I've removed the single-element contained by extensions.

austindrenski commented on src/EFCore.PG/NpgsqlRangeExtensions.cs:
```diff
@@ -0,0 +1,488 @@
+        [Pure]
+        public static bool Overlaps<T>(this NpgsqlRange<T> a, NpgsqlRange<T> b) where T : IComparable<T>
+        {
+            if (a.IsEmpty || b.IsEmpty)
```
Will review

austindrenski commented on src/EFCore.PG/NpgsqlRangeExtensions.cs:
```diff
@@ -0,0 +1,488 @@
+                return false;
+            }
+
+            if (a.LowerBoundInfinite && a.UpperBoundInfinite || b.LowerBoundInfinite && b.UpperBoundInfinite)
```
Will review

austindrenski commented on src/EFCore.PG/NpgsqlRangeExtensions.cs:
```diff
@@ -0,0 +1,488 @@
+        /// <value>true</value> if the first range is strictly to the right of the second; otherwise, <value>false</value>.
+        /// </returns>
+        [Pure]
+        public static bool StrictlyRightOf<T>(this NpgsqlRange<T> a, NpgsqlRange<T> b) where T : IComparable<T>
```
I like it, renamed

austindrenski commented on src/EFCore.PG/NpgsqlRangeExtensions.cs:
```diff
@@ -0,0 +1,488 @@
+        /// <value>true</value> if the ranges are adjacent; otherwise, <value>false</value>.
+        /// </returns>
+        [Pure]
+        public static bool Adjacent<T>(this NpgsqlRange<T> a, NpgsqlRange<T> b) where T : IComparable<T>
```
Renamed

austindrenski commented on src/EFCore.PG/NpgsqlRangeExtensions.cs:
```diff
@@ -0,0 +1,488 @@
+        /// <value>true</value> if the ranges are equal; otherwise, <value>false</value>.
+        /// </returns>
+        [Pure]
+        public static bool Equal<T>(this NpgsqlRange<T> a, NpgsqlRange<T> b) where T : IComparable<T>
```
I had the same thought...I've since gone to intercepting standard ==/Equals calls (as well as comparison symbols)

austindrenski commented on src/EFCore.PG/Query/ExpressionTranslators/Internal/NpgsqlRangeOperatorTranslator.cs:
```diff
@@ -0,0 +1,309 @@
+            RangeOperatorExpression.OperatorType operatorType = RangeOperatorExpression.OperatorType.None;
+
+            if (generic == Equal)
+            {
```
Changed

austindrenski commented on src/EFCore.PG/Query/ExpressionTranslators/Internal/NpgsqlRangeOperatorTranslator.cs:
```diff
@@ -0,0 +1,309 @@
+            {
+                operatorType = RangeOperatorExpression.OperatorType.Equal;
+            }
+            else if (generic == NotEqual)
```
Will review

austindrenski commented on src/EFCore.PG/Query/Expressions/Internal/RangeOperatorExpression.cs:
```diff
@@ -0,0 +1,347 @@
+
+        /// <inheritdoc />
+        [NotNull]
+        public override string ToString()
```
Changed

austindrenski commented on src/EFCore.PG/Query/Expressions/Internal/RangeOperatorExpression.cs:
```diff
@@ -0,0 +1,347 @@
+            switch (operatorType)
+            {
+            case OperatorType.None:
+            {
```
Changed

austindrenski commented on src/EFCore.PG/Query/Sql/Internal/NpgsqlQuerySqlGenerator.cs:
```diff
@@ -1,4 +1,5 @@
 ﻿#region License
+
```
IDE strikes again -- changed

austindrenski commented on src/EFCore.PG/Query/Sql/Internal/NpgsqlQuerySqlGenerator.cs:
```diff
@@ -152,7 +159,7 @@ void GenerateArrayIndex([NotNull] BinaryExpression expression)
                 // bytea cannot be subscripted, but there's get_byte
                 VisitSqlFunction(new SqlFunctionExpression("get_byte", typeof(byte),
-                    new[] { expression.Left, expression.Right }));
+                                                           new[] { expression.Left, expression.Right }));
```
IDE strikes again -- reverted

austindrenski commented on src/EFCore.PG/Query/Sql/Internal/NpgsqlQuerySqlGenerator.cs:
```diff
@@ -162,7 +169,7 @@ void GenerateArrayIndex([NotNull] BinaryExpression expression)
 
                 VisitSqlFunction(new SqlFunctionExpression("substr", typeof(char),
-                    new[] { expression.Left, expression.Right, Expression.Constant(1) }));
+                                                           new[] { expression.Left, expression.Right, Expression.Constant(1) }));
```
IDE strikes again -- reverted

austindrenski commented on src/EFCore.PG/Query/Sql/Internal/NpgsqlQuerySqlGenerator.cs:
```diff
@@ -206,14 +230,17 @@ public Expression VisitRegexMatch([NotNull] RegexMatchExpression regexMatchExpre
             Sql.Append("('(?");
-            if (options.HasFlag(RegexOptions.IgnoreCase)) {
+            if (options.HasFlag(RegexOptions.IgnoreCase))
+            {
```
IDE strikes again -- reverted

austindrenski commented on test/EFCore.PG.FunctionalTests/Query/RangeContainedByNpgsqlQueryTest.cs:
```diff
@@ -0,0 +1,173 @@
+﻿using System.Linq;
```
Moved tests into `RangeQueryNpgsqlTest`

austindrenski commented on test/EFCore.PG.Tests/Storage/NpgsqlTypeMappingTest.cs:
```diff
@@ -44,8 +43,7 @@ public void GenerateSqlLiteral_returns_timestamptz_datetime_literal()
             var offset = TimeZoneInfo.Local.BaseUtcOffset.Hours;
-            var offsetStr = offset < 10 ? $"0{offset}" : offset.ToString();
-            Assert.StartsWith($"TIMESTAMPTZ '1997-12-17 07:37:16+{offsetStr}",
+            Assert.StartsWith($"TIMESTAMPTZ '1997-12-17 07:37:16{offset:+00;-00}",
```
Sure -- reverted

austindrenski commented on src/EFCore.PG/Query/ExpressionTranslators/Internal/NpgsqlRangeOperatorTranslator.cs:
```diff
@@ -0,0 +1,309 @@
+        /// <summary>
+        /// Initializes static resources.
+        /// </summary>
+        static NpgsqlRangeOperatorTranslator()
```
It felt overly-complicated while writing it -- happy to give it another look

roji commented:
OK great, let me know when you want me to take another look.

austindrenski commented:
## Updates
- I've pushed some changes to address the white-space/brace/modifiers/unit test changes. 
  - My IDE settings were a little more aggressive than I realized
  - I'll try to keep an eye on the white-space and blank line diffs going forward

### In-memory implementations:
> I'm not 100% sure of the value of providing in-memory implementations of the range operations. For general, non-PostgreSQL-specific types/operations (like string regex), this makes a lot of sense; it allows the same code to client-evaluate on one database and server-evaluate on another. However, NpgsqlRange is by its nature limited to PostgreSQL only, and I'm not sure where and why someone would want to do these operations in-memory... What do you think?

- I'm in favor of providing in-memory implementations for a handful of reasons:
  1. I have a handful of (personal) use cases where in-memory evaluation would be helpful:
      - e.g. validating user input on the server (web API) before saving back to Postgres
  2. Simplification: 
      - No need to document "this works here, but only if _here_ is evaluated in Postgres"
  3. Makes REPL workflows possible with the Npgsql types:
      - Testing with C# interactive 
      - Might matter more for F# workflows
        - Is there a significant F# community using Npgsql?

- Implicit in the above is that I will rework the in-memory implementations so that they:
  1. Exist
  2. Work correctly
  3. Follow Postgres semantics
     - Review whether any of the semantics change across the Npgsql-supported versions of Postgres 

### `IComparable<T>`:
> We can decide to make IComparable<T> IComparable<T>, but I think we should do it only if we end up providing in-memory implementations of all operations, which I'm not sure about (see above). In any case, the value of this would be somewhat limited, considering we can express comparison with the other operations (greater than, lesser than).

I've reworked the `==`/`Equals(...)` and related comparison symbols to use the built-ins rather than my extension methods. If it works as intended, this should make the comparison operators opt-in on the part of Npgsql. 

Currently, the compiler complains if  `<` is used on `NpgsqlRange<T>`. If we add `IComparable<T>` to `NpgsqlRange<T>`, then `<` will be intercepted and rewritten as a `RangeOperatorExpression`.

Provided my "opt-in" code is working right, this punts the question to the Npgsql project. 

If you think this is worth supporting, I could take a first pass at adding `IComparable<T>` support on the Npgsql side after everything is wrapped up here.

### `range @> range = TRUE`:
> Regarding range @> range = TRUE, see aspnet/EntityFrameworkCore#9143. There's nothing to be done about it at the moment, but there's also no real issue with the = TRUE either.

Great! I wanted to make sure I hadn't missed some code telling EF Core "this operator returns boolean". 

Aesthetically, it would be nice to remove, but looks like the ball is out of our court.

## My next steps

1. Review the method caching and equality in `NpgsqlRangeOperatorTranslator`
2. Review the existing in-memory implementations in `NpgsqlRangeExtensions`
3. Expand unit test coverage

roji commented:
Thanks for all the changes and the reaction speed.

> - I'm in favor of providing in-memory implementations for a handful of reasons:
>  1. I have a handful of (personal) use cases where in-memory evaluation would be helpful:
>      - e.g. validating user input on the server (web API) before saving back to Postgres

This kind of thing is frequently done by settings constraints on the PostgreSQL side (i.e. that a given value is within the range), but I imagine in some scenarios some in-memory operations could be useful.

>  2. Simplification: 
>      - No need to document "this works here, but only if _here_ is evaluated in Postgres"

I don't think that's quite necessary - we're discussing extension methods that come from the Npgsql EF Core provider, it's quite clear that their first purpose is mapping to SQL operations. There are also several other cases where similar operations throw.

In parallel to this there's an implementation of PostgreSQL text search (#315). Short of an in-memory reimplemention of the entire whole text search engine, including linguistic capabilities, methods there will have to throw when executed in memory... Of course the complexity of ranges is nowhere near that of full text search, but this is a good example of where an in-memory implementation simply isn't relevant.

>  3. Makes REPL workflows possible with the Npgsql types:
>      - Testing with C# interactive 
>      - Might matter more for F# workflows
>        - Is there a significant F# community using Npgsql?

OK. Don't forget that you can still do EF Core in REPL, but sure, if you're just playing around with the types and their implementation then it's useful to have in-memory.

> - Implicit in the above is that I will rework the in-memory implementations so that they:
>  1. Exist
>  2. Work correctly
>  3. Follow Postgres semantics
>     - Review whether any of the semantics change across the Npgsql-supported versions of Postgres 

That's an important point - if the in-memory implementation is going to differ in any way from the PostgreSQL implementation, then it's better not to have it at all (better to have a clear exception than a subtle behavior change).

One more objection I haven't made before... Keep in mind that in theory PostgreSQL allows user-defined ranges over any type - including types which on the CLR side may be mapped to non-comparable types. It's true that at this point only the built-in range types are supported (int4range, int8range, numrange, tsrange, tstzrange, daterange), but that may change in the future.

Another even worst possibility is for the ordering of element types to be somehow subtly different between CLR and PostgreSQL, so that the in-memory implementation would return different results than the PostgreSQL one - this would be totally beyond your control.

So there are risks to having an in-memory implementation. That being said, for ranges specifically I think the complexity and risks are lower than other cases, and if you want to fully implement this I'm OK with that.

> I've reworked the ==/Equals(...) and related comparison symbols to use the built-ins rather than my extension methods. If it works as intended, this should make the comparison operators opt-in on the part of Npgsql.

OK, but let's distinguish between equality (==/Equals) and comparison (`IComparable<T>`). For equality, it's important to translate ==, Equals, and also `IEquatable<T>`, which is generic and allows comparing two ranges without boxing (NpgsqlRange is a struct).

> Currently, the compiler complains if < is used on NpgsqlRange<T>. If we add IComparable<T> to NpgsqlRange<T>, then < will be intercepted and rewritten as a RangeOperatorExpression.

There's something I may be missing... If you add `IComparable<T>` in Npgsql, then you have to also provide the in-memory implementation to those methods in Npgsql, don't you?

This can lead to a different way of looking at this... If you really intend to provide all the in-memory implementations as said above, that can be done on the `NpgsqlRange<T>` type itself, rather than via extension methods in the EF Core provider. The EF Core provider would still identify these and translate them to their SQL counterparts, of course. This would provide the benefits of in-memory evaluation even when not using EF Core.

What do you think? If you want to go down this route I think I'd be fine with that. We would probably release this in a patch version of 3.2, to allow EF Core 2.1 to still use Npgsql 3.2 and not force a dependency on 3.3.

Great progress so far!

roji commented on src/EFCore.PG/NpgsqlRangeExtensions.cs:
```diff
@@ -56,52 +55,52 @@ public static class NpgsqlRangeExtensions
 
-            bool testLower = compareLower > 0 || compareLower == 0 && range.LowerBoundIsInclusive;
-            bool testUpper = compareUpper > 0 || compareUpper == 0 && range.UpperBoundIsInclusive;
+            bool lower = a.LowerBoundInfinite || !b.LowerBoundInfinite && compareLower > 0 || compareLower == 0 && a.LowerBoundIsInclusive || !b.LowerBoundIsInclusive;
```
I personally add parentheses in cases like this because I'm incapable of remembering precedence and it's better to be 100% unambiguous...

austindrenski commented:
This most recent commit reworks the unit tests to run through a wider range of cases. 

In doing so, I found that one of the containment cases fails when the containing range (e.g. `[1,10)`) has an inclusive lower bound that is the first incremented value after the non-inclusive lower bound of the contained range (e.g. `(0,5)`).

This query returns true (version 9.6):

```sql
SELECT '[1,10)'::int4range @> '(0,5)'::int4range;
```

This seems like it will be difficult to match with purely generic functions. 

Some thoughts:
- Specific types (e.g. `int`, `datetime`) could use non-generic overloads to allow for the increment.
- Perhaps there's some handling that could be done _within_ `NpgsqlRange<T>` to preempt this situation?

Hoping that I've missed the obvious solution... Any ideas?

roji commented:
I understand. The situation seems even more complex, since whether this containment is true depends on whether the underlying type is discrete or not... In other words, for a range of float/double the above would be false, since '(0,5)' includes 0.5, which isn't included in `[1,5)`...

One way around this without specific checks inside those methods (i.e. `if (typeof(T) == typeof(int)`), cast T to the appropriate value and then increment, since .NET simply doesn't provide a concept of increment generically. That solution doesn't seem too bad, even if it won't take care of the totally of possible user-defined ranges (but that's largely theoretical).

But this seems like a lot of trouble for an in-memory implementation, and there's also a good risk of some edge case discrepancies... I'm even less sure now that it's a good idea to do this.

austindrenski commented:
I agree that the edge case risk is significant...

While I still want to see in-memory support, the known edge cases will take time to get right (and I'm sure that others will be found in the process). 

To keep things moving for EF Core 2.1, I'll split off the in-memory implementations from the translations, and move the current in-memory work into a new branch, `range-operators-in-memory`. The current implementations will be rewritten to throw on `NotSupportedExceptions`.

I'll submit separate PRs as each in-memory operator achieves parity with the Postgres version. I think this will make it easier to spot and contain edge cases during review.

## Next steps
- I'll push a commit in the next few minutes to make the changes described above.
- Then I'll begin adding unit tests to complete coverage of the operator extensions.

roji commented:
> While I still want to see in-memory support, the known edge cases will take time to get right (and I'm sure that others will be found in the process).
> 
> To keep things moving for EF Core 2.1, I'll split off the in-memory implementations from the translations, and move the current in-memory work into a new branch, range-operators-in-memory. The current implementations will be rewritten to throw on NotSupportedExceptions.

I think that's a good idea. The one problem is that if you *are* going to do in-memory, there's the option of putting the methods on the NpgsqlRange type itself - inside Npgsql - rather than as extensions (I mentioned this possibility above). Of course this makes no sense if we end up dropping in-memory.

But in any case, we can start like this and see where it leads us.

roji commented:
@austindrenski don't forget to ping me when you want me to review again.

roji commented:
@austindrenski EF Core 2.1 is around the corner, and I'd like to close whatever outstanding things are left. Any chance we can quickly work on merging this? If you have no time to work on it, let me know and I'll continue the work myself.

roji commented:
@austindrenski, I can see some activity on this PR, but it currently lists 104 commits... This is probably a result of a problematic merge.

The best thing to do would be to take the current Npgsql dev branch as the new base for your PR branch, and cherry-pick your changes on top of it. The PR should show only the commits you've added to implement this functionality. Post back if this is confusing and you need git guidance.

austindrenski commented:
@roji, It looks like I missed your ping from back in March, apologies for that.

As for all the commits... some git guidance on cleaning up the PR would be much appreciated. I seem to have gone a little awry while bringing my branch up to date...

austindrenski commented:
I appreciate the guidance, but I'm still having trouble resolving the conflicts on my end while going through the cherry-picking process.

Is there any harm from to the PR process in rebasing and...manually recreating the changes prior to the force push?

roji commented:
No harm in doing anything - whatever it is you end up doing, in the end you should have a clean local `range-operators` branch which you'll push-force to github in order to update this PR. It doesn't really matter how you do it.

austindrenski commented:
Alright, I think the rebase and change incorporation went through okay. 

Changes since the last round of commits:

- I reworked some of the implementation code and defined `NpgsqlOperatorAttribute` to store translation metadata directly on the `NpgsqlRangeOperatorType` enum. It reduces the number of places that record translation info and (hopefully) simplify future maintenance.

- I also defined `NpgsqlRangeOperatorAttribute` to simplify the identification and mapping of extension methods to operators. This provides a hook to make other methods eligible for translation as range operators. The motivation here is to keep some flexibility for future additions.
  - For example:
    - Migrating extension methods in the EF Core project to instance methods on `NpgsqlRange<T>`.
    - Consumers providing implementations for local evaluation without modifying the translation code.

austindrenski commented:
I took a look at `CustomBinaryExpression` and it _drastically_ reduces the amount of expression code. 

I'm wrapping up a refactor now. If you haven't started yet, I recommend waiting until these commits are pushed.

austindrenski commented:
I finished the refactoring to use `CustomBinaryExpression`, and I think it's in good shape for review.

I tried to keep things extensible with `NpgsqlRangeOperatorTranslator` by providing a hook for other methods to opt into range operator translation. Currently, this works by: 
 1. Declaring an instance of `NpgsqlRangeOperatorAttribute` on the method.
 2. Registering the method with `NpgsqlRangeOperatorTranslator.AddSupportedMethod(MethodInfo)`.

To support these hooks, I moved `NpgsqlRangeOperatorType` and `NpgsqlRangeOperatorAttribute` from `Npgsql.EntityFrameworkCore.PostgreSQL.Query.ExpressionTranslators.Internal` to `Npgsql.EntityFrameworkCore.PostgreSQL.Query.ExpressionTranslators`.

roji commented on src/EFCore.PG/NpgsqlRangeExtensions.cs:
```diff
@@ -0,0 +1,253 @@
+        /// <summary>
+        /// Determines whether a range contains a specified value.
+        /// </summary>
+        /// <param name="range">
```
nit: here and below, why not put the `<param>` and `<typeparam>` in a single line:
```c#
/// <param name="range">The range in which to locate the value</param>
```
This seems to be standard in the .NET / EF Core code bases and reduces lots of useless lines

roji commented on src/EFCore.PG/NpgsqlRangeExtensions.cs:
```diff
@@ -0,0 +1,253 @@
+        /// <value>true</value> if the range contains the specified value; otherwise, <value>false</value>.
+        /// </returns>
+        [NpgsqlRangeOperator(NpgsqlRangeOperatorType.Contains)]
+        public static bool Contains<T>(this NpgsqlRange<T> range, T value) where T : IComparable<T> => throw new NotImplementedException();
```
I think this should throw `NotSupportException`, not `NotImplementedException`.

roji commented on test/EFCore.PG.FunctionalTests/Query/RangeQueryNpgsqlTest.cs:
```diff
@@ -0,0 +1,514 @@
+                           .Where(x => !x.Range.Contains(range))
+                           .ToArray();
+
+                Assert.Contains("WHERE NOT (x.\"Range\" @> @__range_0 = TRUE)", Fixture.TestSqlLoggerFactory.Sql);
```
Nit: Refactor this out to a single `AssertContainsSql()` which can be called from all the tests.

roji commented on src/EFCore.PG/Query/ExpressionTranslators/Internal/NpgsqlRangeOperatorTranslator.cs:
```diff
@@ -0,0 +1,189 @@
+    /// By default, this class supports translation for methods declared on <see cref="NpgsqlRangeExtensions"/>.
+    /// Additional methods can be supported so long as they declare a <see cref="NpgsqlRangeOperatorAttribute"/>.
+    /// </remarks>
+    public class NpgsqlRangeOperatorTranslator : IMethodCallTranslator
```
Please rename to NpgsqlRangeTranslator to align with the naming scheme of other translators.

roji commented on src/EFCore.PG/Query/ExpressionTranslators/Internal/NpgsqlBinaryOperatorAttribute.cs:
```diff
@@ -0,0 +1,81 @@
+            Check.NotNull(right, nameof(right));
+
+            // TODO: will the left type arguments always be valid?
+            Type type = ReturnType.IsGenericType ? ReturnType.MakeGenericType(left.Type.GetGenericArguments()) : ReturnType;
```
nit: please use `var` everywhere

roji commented on src/EFCore.PG/Query/ExpressionTranslators/Internal/NpgsqlBinaryOperatorAttribute.cs:
```diff
@@ -0,0 +1,81 @@
+            Check.NotNull(left, nameof(left));
+            Check.NotNull(right, nameof(right));
+
+            // TODO: will the left type arguments always be valid?
```
See my general comment about the architecture. Simply moving all the logic into the translator makes this go away, as you'd be translating each operation separately and can make this decision on a per-operation basis.

austindrenski commented:
@roji Thanks for the prompt review. I've pushed a commit addressing your comments, and it's ready for another round of review.

> There's simply a great number of different classes to support this mechanism

> I think it's healthier to have one simple way to translate, rather than have a special attribute-based mechanism for Npgsql-defined extension methods, and another mechanism for the others.

Those are very fair points. Perhaps too much time with ASP.NET Core has made every problem look well-suited for an attribute-based solution. 

Incorporating your comments eliminated a bunch of complexity and removed 3 extra files.

Appreciate all the feedback!

roji commented on src/EFCore.PG/Query/ExpressionTranslators/Internal/NpgsqlRangeTranslator.cs:
```diff
@@ -0,0 +1,110 @@
+        /// The expression if successful; otherwise, null.
+        /// </returns>
+        [CanBeNull]
+        static Expression TryTranslateOperator([NotNull] MethodCallExpression expression)
```
Where not merge `TryTranslateOperator` with `Translate` above, to just have the entire switch in `Translate()`? The separation doesn't seem to add any value. In addition, methods start with Try in C# normally return a bool (success/failure) and have an `out` var, so this method isn't really following standard naming.

roji commented on src/EFCore.PG/Query/ExpressionTranslators/Internal/NpgsqlRangeTranslator.cs:
```diff
@@ -0,0 +1,110 @@
+        /// A <see cref="CustomBinaryExpression"/>.
+        /// </returns>
+        [NotNull]
+        static Expression MakeBinaryExpression([NotNull] MethodCallExpression expression, [NotNull] string symbol, [NotNull] Type returnType) =>
```
Similar to the above, why not just instantiate `CustomBinaryExpression` inside the switch, rather than calling this method which does it? There's almost zero value added here.

roji commented on test/EFCore.PG.FunctionalTests/Query/RangeQueryNpgsqlFixture.cs:
```diff
@@ -0,0 +1,158 @@
+    /// <summary>
+    /// Represents a fixture suitable for testing range operators.
+    /// </summary>
+    public class RangeQueryNpgsqlFixture : IDisposable
```
In EF Core the standard now seems to be to include the fixture as an inner class of the test suite class. This makes sense since the fixture is only used within its test suite, and this way we have one file per test suite - easier to navigate the ever-growing list of test suites. Let's do this here as well.
